File: 108. Convert Sorted Array to Binary Search Tree/108.ConvertSortedArraytoBinarySearchTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode sortedArrayToBST(int[] nums) {
        // Recursive helper function to build the BST
        return buildBST(nums, 0, nums.length - 1);
    }

    private TreeNode buildBST(int[] nums, int left, int right) {
        // Base case: if the range is invalid, return null
        if (left > right) return null;

        // Find the middle element to make it the root
        int mid = (left + right) / 2;
        TreeNode root = new TreeNode(nums[mid]);

        // Recursively build left and right subtrees
        root.left = buildBST(nums, left, mid - 1);
        root.right = buildBST(nums, mid + 1, right);

        // Return the root node
        return root;
    }
}

File: 108. Convert Sorted Array to Binary Search Tree/108.ConvertSortedArraytoBinarySearchTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val === undefined ? 0 : val);
 *     this.left = (left === undefined ? null : left);
 *     this.right = (right === undefined ? null : right);
 * }
 */

/**
 * Converts a sorted array into a height-balanced Binary Search Tree (BST).
 * @param {number[]} nums - The sorted array.
 * @return {TreeNode} - The root node of the height-balanced BST.
 */
var sortedArrayToBST = function (nums) {
  // Helper function to build the BST recursively
  function buildBST(left, right) {
    // Base case: if left index exceeds right, there are no elements to include
    if (left > right) return null;

    // Middle element is chosen as the root for the balanced BST
    const mid = Math.floor((left + right) / 2);

    // Create the root node with the middle element
    const root = new TreeNode(nums[mid]);

    // Recursively build the left and right subtrees
    root.left = buildBST(left, mid - 1); // Left half
    root.right = buildBST(mid + 1, right); // Right half

    // Return the current root node
    return root;
  }

  // Start the recursive process
  return buildBST(0, nums.length - 1);
};
